### PR TITLE
fix(gitlab): gitLabIgnoreApprovals works with multiple approval rules

### PR DESCRIPTION
**Pull request by t-kulmburg**
## Changes

Currently `gitLabIgnoreApprovals` sets the `approvals_required` parameter to 0 only for the default `any_approver` rule.
If additional custom approval rules are in place in the project configuration, automerge is not working anymore since GitLab 1.16 because of this: https://gitlab.com/gitlab-org/gitlab/-/issues/388140

## Context

Simpler fix compared to https://github.com/renovatebot/renovate/pull/22306, without adding an additional config option.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by cjpearson** [comment]
I also don't see the use case for the config option and think it's fine to remove. My preference would be to remove approval rules instead of setting the approver count to zero, but I don't know if that makes a difference in practice. I just think it makes renovate's behavior more obvious and I wasn't sure if resetting the count would work for all types of approval rules.

Also, are we sure every MR would have at least one `any_approver` rule?

File: lib/modules/platform/gitlab/index.ts [hunk 1]
```diff
@@ -543,9 +543,11 @@ async function ignoreApprovals(pr: number): Promise<void> {
     );
 
     if (existingAnyApproverRule) {
-      await gitlabApi.putJson(`${url}/${existingAnyApproverRule.id}`, {
-        body: { ...existingAnyApproverRule, approvals_required: 0 },
-      });
+      for (const rule of rules) {
+        await gitlabApi.putJson(`${url}/${rule.id}`, {
+          body: { ...existingAnyApproverRule, approvals_required: 0 },
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:548]: this seems wrong, you're overwriting the rule with an any approver rule, so you would finally have multiple of them, which isn't allowed on gitlab.

i think we should simply delete all rules when `gitLabIgnoreApprovals=true`

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/platform/gitlab/index.ts [hunk 2]
```diff
@@ -538,20 +538,28 @@ async function ignoreApprovals(pr: number): Promise<void> {
...
       ({ rule_type }) => rule_type === 'any_approver'
     );
+    const existingRegularApproverRules = rules?.filter(
+      ({ rule_type, name }) => rule_type !== 'any_approver' && name !== ruleName
+    );
+
+    await Promise.all(
+      existingRegularApproverRules?.map(async (rule) => {
+        await gitlabApi.deleteJson(`${url}/${rule.id}`);
+      })
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:553]: use `all` from `util/promises`to limit concurrency.

https://github.com/renovatebot/renovate/blob/45c2d13f8a1dfffe432e3cc02892119cca7131ef/lib/util/promises.ts#L38-L52

File: lib/modules/platform/gitlab/index.ts [hunk 3]
```diff
@@ -538,9 +538,20 @@ async function ignoreApprovals(pr: number): Promise<void> {
...
+
     const existingAnyApproverRule = rules?.find(
       ({ rule_type }) => rule_type === 'any_approver'
     );
+    const existingRegularApproverRules = rules?.filter(
+      ({ rule_type, name }) => rule_type !== 'any_approver' && name !== ruleName
+    );
+
+    await Promise.all(
+      existingRegularApproverRules?.map(async (rule) => {
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:551]: `Promise.all(undefined)` rejects, so add a condition to only do `if existingRegularApproverRules?.length`

File: lib/modules/platform/gitlab/index.ts [hunk 4]
```diff
@@ -538,9 +538,22 @@ async function ignoreApprovals(pr: number): Promise<void> {
...
+    const ruleName = 'renovateIgnoreApprovals';
+
     const existingAnyApproverRule = rules?.find(
       ({ rule_type }) => rule_type === 'any_approver'
     );
+    const existingRegularApproverRules = rules?.filter(
+      ({ rule_type, name }) => rule_type !== 'any_approver' && name !== ruleName
+    );
+
+    if (existingRegularApproverRules.length) {
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:550]:
```suggestion
    if (existingRegularApproverRules?.length) {
```
will fix cannot read length of undefined 🙃

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
like to have some second 👀

**Pull request merged**